laws/shared/src/main/scala/algebra/laws/RingLaws.scala

```diff
@@ -201,6 +201,57 @@ trait RingLaws[A] extends GroupLaws[A] { self =>
     parents = Seq(ring, commutativeRig, commutativeRng)
   )
 
+  def gcdRing(implicit A: GCDRing[A]) = RingProperties.fromParent(
+    name = "gcd domain",
+    parent = commutativeRing,
+    "gcd/lcm" -> forAll { (x: A, y: A) =>
+      val d = A.gcd(x, y)
+      val m = A.lcm(x, y)
+      A.times(x, y) ?== A.times(d, m)
+    },
+    "gcd is commutative" -> forAll { (x: A, y: A) =>
+      A.gcd(x, y) ?== A.gcd(y, x)
+    },
+    "lcm is commutative" -> forAll { (x: A, y: A) =>
+      A.lcm(x, y) ?== A.lcm(y, x)
+    },
+    "gcd(0, 0)" -> (A.gcd(A.zero, A.zero) ?== A.zero),
+    "lcm(0, 0) === 0" -> (A.lcm(A.zero, A.zero) ?== A.zero),
+    "lcm(x, 0) === 0" -> forAll { (x: A) => A.lcm(x, A.zero) ?== A.zero }
+  )
+
+  def euclideanRing(implicit A: EuclideanRing[A]) = RingProperties.fromParent(
+    name = "euclidean ring",
+    parent = gcdRing,
+    "euclidean division rule" -> forAll { (x: A, y: A) =>
+      pred(y) ==> {
+        val (q, r) = A.equotmod(x, y)
+        x ?== A.plus(A.times(y, q), r)
+      }
+    },
+    "equot" -> forAll { (x: A, y: A) =>
+      pred(y) ==> {
+        A.equotmod(x, y)._1 ?== A.equot(x, y)
+      }
+    },
+    "emod" -> forAll { (x: A, y: A) =>
+      pred(y) ==> {
+        A.equotmod(x, y)._2 ?== A.emod(x, y)
+      }
+    },
+    "euclidean function" -> forAll { (x: A, y: A) =>
+      pred(y) ==> {
+        val (_, r) = A.equotmod(x, y)
+        A.isZero(r) || (A.euclideanFunction(r) < A.euclideanFunction(y))
+      }
+    },
+    "submultiplicative function" -> forAll { (x: A, y: A) =>
```

Comment:
I didn't see this comment in trait. Could we add it so people get more intuition for the euclideanFunction?

Reply:
Yes!